File: wellnest/wellnest/page/ops_dashboard/ops_dashboard.js
frappe.pages['ops-dashboard'].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Ops Dashboard',
        single_column: true,
    });

    var css = `
        .filter-input {
            width: auto;
            display: inline-block;
            margin-right: 10px;
            max-width: 250px;
            font-size: 14px;
        }

        .apply-button, .remove-filter-button {
            display: inline-block;
            font-size: 14px;
            margin-top: 5px;
        }

        .filters {
            display: flex;
            flex-wrap: wrap;
            align-items: center;
            gap: 10px;
        }

        .filters h4 {
            margin-right: 10px;
            font-size: 16px;
        }

        .status-engaged {
            color: green;
            font-size: 16px;
            display: inline-block;
            margin-left: 10px;
            font-weight: bold;
        }

        .status-not-engaged {
            color: red;
            font-size: 16px;
            display: inline-block;
            margin-left: 10px;
            font-weight: bold;
        }

        .status-dot {
            border-radius: 50%;
            width: 10px;
            height: 10px;
            display: inline-block;
            margin-left: 5px;
        }

        .status-engaged-dot {
            background-color: green;
        }

        .status-not-engaged-dot {
            background-color: red;
        }
    `;

    var style = $('<style>').text(css);
    $('head').append(style);

    let current_page = 1;
    const items_per_page = 10;
    let current_filters = {};

    function load_dashboard_data(filters = {}, page_number = 1) {
        const offset = (page_number - 1) * items_per_page;

        
    
        frappe.call({
            method: 'wellnest.wellnest.page.ops_dashboard.ops_dashboard.get_dashboard_data',
            args: { filters: JSON.stringify(filters), limit: items_per_page, offset: offset },
            callback: function (response) {
                var data = response.message;
    
                page.main.empty();
    
                create_filters(page.main, load_dashboard_data, filters);
    
                render_table(page.main, "Caregivers", data.caregivers, [
                    "caregiver_name", "caregiver_type", "last_engagement", "status", "total_accrued_amount", "invoice_raised", "due_amount", "paid_amount", "engagement_id"
                ]);
    
                render_table(page.main, "Customers", data.customers, [
                    "customer_name",
                    "status",
                    "last_engagement",
                    "invoice_raised",
                    "invoice_raised_date",
                    "outstanding_total",
                    "paid_amount",
                    "assigned_caregivers"  
                ]);
    
                create_pagination_controls(page.main, page_number, data.has_more, filters);
            }
        });
    }
    

    function create_filters(parent, on_apply_filters, filters) {
        var filter_section = $('<div class="filters" style="margin-bottom: 20px;">');
        var caregiver_filter = $('<input type="text" placeholder="Search Caregiver" class="form-control filter-input">');
        var customer_filter = $('<input type="text" placeholder="Search Customer" class="form-control filter-input">');
        var apply_button = $('<button class="btn btn-primary apply-button">Apply Filters</button>');
        var remove_filter_button = $('<button class="btn btn-secondary remove-filter-button">Remove Filter</button>');

        caregiver_filter.val(filters.caregiver_name || "");
        customer_filter.val(filters.customer_name || "");

        apply_button.on('click', function () {
            var filters = {
                caregiver_name: caregiver_filter.val(),
                customer_name: customer_filter.val()
            };
            current_filters = filters;
            on_apply_filters(filters, 1);
        });

        remove_filter_button.on('click', function () {
            caregiver_filter.val("");
            customer_filter.val("");
            current_filters = {};
            on_apply_filters({}, 1);
        });

        filter_section.append(caregiver_filter, customer_filter, apply_button, remove_filter_button);
        parent.append(filter_section);
    }

    function render_table(parent, title, data, columns) {
        data.sort((a, b) => {
            // Sort by status first: 'Engaged' should be higher priority than 'Not Engaged'
            if (a.status === "Engaged" && b.status !== "Engaged") return -1;
            if (a.status !== "Engaged" && b.status === "Engaged") return 1;
        
            // If both have the same status, sort by last engagement date in descending order
            const a_last_engagement = a.last_engagement ? new Date(a.last_engagement) : 0;
            const b_last_engagement = b.last_engagement ? new Date(b.last_engagement) : 0;
        
            return b_last_engagement - a_last_engagement; // Sort latest engagement first
        });
        

    var table_html = `
        <h3>${title}</h3>
        <table class="table table-bordered table-striped">
            <thead>
                <tr>
                    ${columns.map(col => `<th>${capitalize(col.replace('_', ' '))}</th>`).join('')}
                </tr>
            </thead>
            <tbody>
                ${data.map(row => {
                    return `
                        <tr>
                            ${columns.map(col => {
                                if (col === "caregiver_name") {
                                    return `<td><a href="javascript:void(0)" class="caregiver-link" data-caregiver="${row[col]}">${row[col]}</a></td>`;
                                } else if (col === "status") {
                                    return `<td>${row[col]} <span class="status-dot ${row[col] === 'Engaged' ? 'status-engaged-dot' : 'status-not-engaged-dot'}"></span></td>`;
                                } else if (col === "last_engagement") {
                                    return `<td>${row[col] ? new Date(row[col]).toLocaleDateString() : ''}</td>`;
                                } else if (col === "invoice_raised_date") {
                                    return `<td>${row[col] ? new Date(row[col]).toLocaleDateString() : 'No Date'}</td>`;
                                } else if (col === "assigned_caregivers") {
                                    const caregivers = row["assigned_caregivers"] ? row["assigned_caregivers"].split(', ') : "No caregivers assigned";
                                    return `<td>${caregivers}</td>`;
                                } else {
                                    return `<td>${row[col] || ''}</td>`;
                                }
                            }).join('')}
                        </tr>
                    `;
                }).join('')}
            </tbody>
        </table>
    `;
    parent.append(table_html);

    // Prevent double event binding
    parent.find('.caregiver-link').off('click').on('click', function () {
        const caregiver_name = $(this).data('caregiver');
        show_caregiver_popup(caregiver_name);
    });
}

    

    function show_caregiver_popup(caregiver_name) {
        frappe.call({
            method: 'wellnest.wellnest.page.ops_dashboard.ops_dashboard.get_engagements_for_caregiver',
            args: { caregiver_name: caregiver_name },
            callback: function (response) {
                const engagements = response.message.engagements || [];
                const invoices = response.message.invoices ? [response.message.invoices] : [];
                const payments = response.message.payments ? [response.message.payments] : [];
    
                // Prepare Engagements Table
                const engagementRows = engagements.length
                    ? engagements.map(e => `
                        <tr>
                            <td>${e.engagement_id}</td>
                            <td>${e.start_date ? new Date(e.start_date).toLocaleDateString() : '-'}</td>
                            <td>${e.end_date ? new Date(e.end_date).toLocaleDateString() : '-'}</td>
                        </tr>
                    `).join('')
                    : `<tr><td colspan="3">No engagements found</td></tr>`;
    
                // Prepare Invoices Table
                const invoiceRows = invoices.length
                    ? invoices.map(i => `
                        <tr>
                            <td>${i.outstanding_amount}</td>
                        </tr>
                    `).join('')
                    : `<tr><td>No invoices found</td></tr>`;
    
                // Prepare Payments Table
                const paymentRows = payments.length
                    ? payments.map(p => `
                        <tr>
                            <td>${p.amount_paid}</td>
                        </tr>
                    `).join('')
                    : `<tr><td>No payments found</td></tr>`;
    
                const popupHtml = `
                    <div>
                        <h3>Caregiver Details</h3>
                        <p><strong>Name:</strong> ${caregiver_name}</p>
    
                        <h4>Engagements</h4>
                        <table class="table table-bordered table-striped">
                            <thead>
                                <tr>
                                    <th>Engagement ID</th>
                                    <th>Start Date</th>
                                    <th>End Date</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${engagementRows}
                            </tbody>
                        </table>
    
                        <h4>Generated Invoices</h4>
                        <table class="table table-bordered table-striped">
                            <thead>
                                <tr>
                                    <th>Outstanding Amount</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${invoiceRows}
                            </tbody>
                        </table>
    
                        <h4>Paid Out</h4>
                        <table class="table table-bordered table-striped">
                            <thead>
                                <tr>
                                    <th>Amount Paid</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${paymentRows}
                            </tbody>
                        </table>
                    </div>
                `;
    
                // Display popup with the caregiver details
                frappe.msgprint({
                    title: `Details for ${caregiver_name}`,
                    indicator: 'green',
                    message: `<div>${popupHtml}</div>`, // Wrap content to avoid re-adding content
                });
            },
        });
    }

    function create_pagination_controls(parent, page_number, has_more, filters) {
        const pagination = $('<div class="pagination-controls">');
        if (page_number > 1) {
            const prevButton = $('<button class="btn btn-secondary">Previous</button>');
            prevButton.on('click', function () {
                load_dashboard_data(filters, page_number - 1);
            });
            pagination.append(prevButton);
        }

        if (has_more) {
            const nextButton = $('<button class="btn btn-secondary">Next</button>');
            nextButton.on('click', function () {
                load_dashboard_data(filters, page_number + 1);
            });
            pagination.append(nextButton);
        }

        parent.append(pagination);
    }

    load_dashboard_data();
};

function capitalize(s) {
    return s.charAt(0).toUpperCase() + s.slice(1);
}

File: wellnest/wellnest/page/ops_dashboard/ops_dashboard.py
import frappe
from frappe import _

@frappe.whitelist()
def get_dashboard_data(filters=None, limit=10, offset=0):
    """
    API method to fetch caregiver and customer data for Ops Dashboard.
    Filters can include 'caregiver_name' and 'customer_name'.
    """
    filters = frappe.parse_json(filters) if filters else {}
    limit = int(limit)
    offset = int(offset)

    caregivers = fetch_caregivers(filters.get('caregiver_name'), limit, offset)
    customers = fetch_customers(filters.get('customer_name'), limit, offset)

    # Fetch caregiver-specific amounts
    total_accrued_amounts = fetch_total_accrued_amount()
    invoice_raised_data = fetch_invoice_raised_amount()
    due_amount_data = fetch_due_amount()
    paid_amount_data = fetch_paid_amount()

    # Optimize lookups using dictionaries
    total_accrued_amounts_dict = {item['caregiver_name']: item['total_accrued_amount'] for item in total_accrued_amounts}
    invoice_raised_data_dict = {item['caregiver_name']: item['invoice_raised'] for item in invoice_raised_data}
    due_amount_data_dict = {item['caregiver_name']: item['due_amount'] for item in due_amount_data}
    paid_amount_data_dict = {item['caregiver_name']: item['paid_amount'] for item in paid_amount_data}

    # Merge caregiver data
    for caregiver in caregivers:
        caregiver_name = caregiver['caregiver_name']
        caregiver['total_accrued_amount'] = total_accrued_amounts_dict.get(caregiver_name, 0)
        caregiver['invoice_raised'] = invoice_raised_data_dict.get(caregiver_name, 0)
        caregiver['due_amount'] = due_amount_data_dict.get(caregiver_name, 0)
        caregiver['paid_amount'] = paid_amount_data_dict.get(caregiver_name, 0)

    # Fetch customer-specific amounts
    invoice_totals = {item['customer_name']: item['invoice_raised'] for item in fetch_customer_invoice_raised()}
    paid_totals = {item['customer_name']: item['paid_amount'] for item in fetch_customer_paid_amount()}
    outstanding_totals = {item['customer_name']: item['outstanding_total'] for item in fetch_customer_outstanding()}

    # Merge invoice raised dates
    invoice_raised_dates_dict = {item['customer_name']: item['invoice_raised_date'] for item in fetch_invoice_raised_dates()}

    # Merge customer data
    for customer in customers:
        customer_name = customer['customer_name']
        customer['invoice_total'] = invoice_totals.get(customer_name, 0)
        customer['paid_total'] = paid_totals.get(customer_name, 0)
        customer['outstanding_total'] = outstanding_totals.get(customer_name, 0)
        customer['invoice_raised_date'] = invoice_raised_dates_dict.get(customer_name, None)

    has_more_caregivers = len(caregivers) == limit
    has_more_customers = len(customers) == limit
    

    return {
    "caregivers": caregivers,
    "customers": customers,
    "has_more": has_more_caregivers or has_more_customers  # Combine both flags
}


def fetch_invoice_raised_dates():
    query = """
        SELECT 
            si.customer AS customer_name, 
            MAX(si.posting_date) AS invoice_raised_date
        FROM `tabSales Invoice` si
        WHERE si.customer IS NOT NULL 
            AND si.docstatus = 1  -- Only submitted invoices
        GROUP BY si.customer
    """
    invoice_raised_dates = frappe.db.sql(query, as_dict=True)
    return invoice_raised_dates

@frappe.whitelist()
def get_assigned_caregivers(customer_name):
    try:
        frappe.logger().info(f"Fetching assigned caregivers for customer: {customer_name}")

        # Query to fetch caregiver names assigned to the customer's engagements
        assigned_caregivers = frappe.db.sql("""
            SELECT 
                caregiver.full_name AS caregiver_name
            FROM `tabEngagement` e
            INNER JOIN `tabEngagement Caregiver` ec ON ec.parent = e.name
            INNER JOIN `tabCaregiver` caregiver ON ec.caregiver = caregiver.name
            WHERE e.customer = %s
            ORDER BY caregiver.full_name
        """, (customer_name,), as_dict=True)

        frappe.logger().info(f"Assigned caregivers fetched: {assigned_caregivers}")

        # Return only the names of the assigned caregivers
        return [caregiver['caregiver_name'] for caregiver in assigned_caregivers] or []
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Ops Dashboard: get_assigned_caregivers")
        frappe.throw(_("An error occurred while fetching assigned caregivers for the customer."))


# Caregiver Functions
def fetch_caregivers(caregiver_name_filter, limit, offset):
    conditions = "1=1"
    params = {"limit": limit, "offset": offset}

    if caregiver_name_filter:
        conditions += " AND caregiver.full_name LIKE %(caregiver_name)s"
        params["caregiver_name"] = f"%{caregiver_name_filter}%"

    query = f"""
        SELECT 
            caregiver.full_name AS caregiver_name,
            caregiver.caregiver_type,
            MAX(ec.start_date) AS last_engagement,
            CASE 
                WHEN MAX(ec.end_date) >= CURDATE() THEN 'Engaged'
                ELSE 'Not Engaged'
            END AS status,
            (
                SELECT ec.parent
                FROM `tabEngagement Caregiver` ec
                WHERE ec.caregiver = caregiver.name
                ORDER BY ec.start_date DESC, ec.creation DESC
                LIMIT 1
            ) AS engagement_id
        FROM `tabCaregiver` caregiver
        LEFT JOIN `tabEngagement Caregiver` ec ON ec.caregiver = caregiver.name
        WHERE {conditions}
        GROUP BY caregiver.full_name, caregiver.caregiver_type
        LIMIT %(limit)s OFFSET %(offset)s
    """
    return frappe.db.sql(query, params, as_dict=True)

def fetch_total_accrued_amount():
    query = """
        SELECT 
            caregiver.full_name AS caregiver_name,
            SUM(
                (DATEDIFF(ec.end_date, ec.start_date) + 1) * ec.daily_rate
            ) AS total_accrued_amount
        FROM `tabEngagement Caregiver` ec
        LEFT JOIN `tabCaregiver` caregiver ON ec.caregiver = caregiver.name
        WHERE ec.start_date IS NOT NULL AND ec.end_date IS NOT NULL
        GROUP BY caregiver.full_name
    """
    return frappe.db.sql(query, as_dict=True)

def fetch_invoice_raised_amount():
    query = """
        SELECT 
            supplier_name AS caregiver_name,
            SUM(grand_total) AS invoice_raised
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1
        AND status IN ('Unpaid', 'Overdue')
        GROUP BY supplier_name
    """
    return frappe.db.sql(query, as_dict=True)

def fetch_due_amount():
    query = """
        SELECT 
            caregiver_name,
            SUM(total_accrued_amount) - SUM(paid_amount) AS due_amount
        FROM (
            -- Fetch total accrued amount per caregiver
            SELECT 
                caregiver.full_name AS caregiver_name,
                SUM((DATEDIFF(ec.end_date, ec.start_date) + 1) * ec.daily_rate) AS total_accrued_amount,
                0 AS paid_amount
            FROM `tabEngagement Caregiver` ec
            LEFT JOIN `tabCaregiver` caregiver ON ec.caregiver = caregiver.name
            WHERE ec.start_date IS NOT NULL AND ec.end_date IS NOT NULL
            GROUP BY caregiver.full_name

            UNION ALL

            -- Fetch paid amount from purchase invoices
            SELECT 
                pi.supplier_name AS caregiver_name,
                0 AS total_accrued_amount,
                SUM(pi.grand_total) - SUM(pi.outstanding_amount) AS paid_amount
            FROM `tabPurchase Invoice` pi
            WHERE pi.docstatus = 1
            AND pi.status = 'Paid'
            GROUP BY pi.supplier_name
        ) AS combined
        GROUP BY caregiver_name
    """
    return frappe.db.sql(query, as_dict=True)

def fetch_paid_amount():
    query = """
        SELECT 
            supplier_name AS caregiver_name,
            SUM(grand_total) - SUM(outstanding_amount) AS paid_amount
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1
        AND status = 'Paid'
        GROUP BY supplier_name
    """
    return frappe.db.sql(query, as_dict=True)



# fetching Customer Details
def fetch_customers(customer_name_filter, limit, offset):
    conditions = "1=1"
    params = {"limit": limit, "offset": offset}

    if customer_name_filter:
        conditions += " AND customer.name LIKE %(customer_name)s"
        params["customer_name"] = f"%{customer_name_filter}%"

    query = f"""
        SELECT 
            customer.name AS customer_name,
            (
                SELECT MAX(e.end_date)  
                FROM `tabEngagement` e
                WHERE e.customer = customer.name
            ) AS last_engagement,
            CASE
                WHEN EXISTS (
                    SELECT 1
                    FROM `tabEngagement` e
                    WHERE e.customer = customer.name
                    AND e.end_date IS NULL
                ) THEN 'Engaged'
                ELSE 'Not Engaged'
            END AS status,
            (
                SELECT SUM(invoice.grand_total)
                FROM `tabSales Invoice` invoice
                WHERE invoice.customer = customer.name
                AND invoice.status IN ('Unpaid', 'Overdue')
            ) AS invoice_raised,
            (
                SELECT SUM(invoice.grand_total) - SUM(invoice.outstanding_amount)
                FROM `tabSales Invoice` invoice
                WHERE invoice.customer = customer.name
                AND invoice.status = 'Paid'
            ) AS paid_amount,
            (
                SELECT SUM(invoice.outstanding_amount)
                FROM `tabSales Invoice` invoice
                WHERE invoice.customer = customer.name
                AND invoice.status IN ('Unpaid', 'Overdue')
            ) AS outstanding_total,
            (
                SELECT GROUP_CONCAT(DISTINCT caregiver.full_name SEPARATOR ', ')
                FROM `tabEngagement` e
                INNER JOIN `tabEngagement Caregiver` ec ON ec.parent = e.name
                INNER JOIN `tabCaregiver` caregiver ON ec.caregiver = caregiver.name
                WHERE e.customer = customer.name
            ) AS assigned_caregivers
        FROM `tabCustomer` customer
        WHERE {conditions}
        LIMIT %(limit)s OFFSET %(offset)s
    """
    return frappe.db.sql(query, params, as_dict=True)



def fetch_customer_invoice_raised():
    query = """
        SELECT 
            customer AS customer_name,
            SUM(grand_total) AS invoice_raised
        FROM `tabSales Invoice`
        WHERE docstatus = 1
        AND status IN ('Unpaid', 'Overdue') -- Include only Unpaid and Overdue statuses
        GROUP BY customer
    """
    return frappe.db.sql(query, as_dict=True)


def fetch_customer_paid_amount():
    query = """
        SELECT 
            customer AS customer_name,
            SUM(grand_total) - SUM(outstanding_amount) AS paid_amount
        FROM `tabSales Invoice`
        WHERE docstatus = 1
        AND status = 'Paid'
        GROUP BY customer
    """
    return frappe.db.sql(query, as_dict=True)

def fetch_customer_outstanding():
    query = """
        SELECT 
            customer AS customer_name,
            SUM(outstanding_amount) AS outstanding_total
        FROM `tabSales Invoice`
        WHERE docstatus = 1
        AND status IN ('Unpaid', 'Overdue')
        GROUP BY customer
    """
    return frappe.db.sql(query, as_dict=True)



@frappe.whitelist()
def get_engagements_for_caregiver(caregiver_name):
    try:
        frappe.logger().info(f"Fetching engagements, invoices, and payments for caregiver (full name): {caregiver_name}")

        # Engagements Query (unchanged)
        engagements = frappe.db.sql("""
            SELECT 
                e.name AS engagement_id,
                ec.start_date,
                ec.end_date
            FROM `tabEngagement` e
            INNER JOIN `tabEngagement Caregiver` ec ON e.name = ec.parent
            INNER JOIN `tabCaregiver` c ON ec.caregiver = c.name
            WHERE c.full_name = %s
            ORDER BY e.creation DESC
        """, (caregiver_name,), as_dict=True)

        frappe.logger().info(f"Engagements fetched: {engagements}")

        # Invoice Query (grouped by supplier_name)
        invoices = frappe.db.sql("""
            SELECT 
                supplier_name AS caregiver_name,
                SUM(outstanding_amount) AS outstanding_amount
            FROM `tabPurchase Invoice`
            WHERE docstatus = 1
            AND status IN ('Unpaid', 'Overdue')
            AND supplier_name = %s
            GROUP BY supplier_name
        """, (caregiver_name,), as_dict=True)

        frappe.logger().info(f"Invoices fetched: {invoices}")

        # Payments Query (grouped by supplier_name)
        payments = frappe.db.sql("""
            SELECT 
                supplier_name AS caregiver_name,
                SUM(grand_total) AS amount_paid
            FROM `tabPurchase Invoice`
            WHERE docstatus = 1
            AND status = 'Paid'
            AND supplier_name = %s
            GROUP BY supplier_name
        """, (caregiver_name,), as_dict=True)

        frappe.logger().info(f"Payments fetched: {payments}")

        # Return data
        return {
            "engagements": engagements or [],
            "invoices": invoices[0] if invoices else {"caregiver_name": caregiver_name, "outstanding_amount": 0},
            "payments": payments[0] if payments else {"caregiver_name": caregiver_name, "amount_paid": 0},
        }

    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Ops Dashboard: get_engagements_for_caregiver")
        frappe.throw(_("An error occurred while fetching engagements, invoices, and payments for the caregiver."))

